### PR TITLE
fix(common): preserve Map's ordered keys with keyvalue pipe

### PR DESCRIPTION
**Pull request by ravishivt**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #31420

## What is the new behavior?

Preserve Map's ordered keys with keyvalue pipe.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

Technically, this is a breaking change but I'd argue that the original design was unexpected behavior (adjusting sort of a Map input).

### CODE REVIEW

File: packages/common/src/pipes/keyvalue_pipe.ts [hunk 1]
```diff
@@ -64,11 +68,14 @@ export class KeyValuePipe implements PipeTransform {
...
       input: null|{[key: string]: V, [key: number]: V}|Map<K, V>,
-      compareFn: (a: KeyValue<K, V>, b: KeyValue<K, V>) => number = defaultComparator):
-      Array<KeyValue<K, V>>|null {
+      compareFn?: (a: KeyValue<K, V>, b: KeyValue<K, V>) => number): Array<KeyValue<K, V>>|null {
     if (!input || (!(input instanceof Map) && typeof input !== 'object')) {
       return null;
     }
+    // Assign the default comparison for non-Map inputs.
+    if (!(input instanceof Map)) {
+      compareFn = defaultComparator;
```
Comment (Airblader) [inline packages/common/src/pipes/keyvalue_pipe.ts]: Now you're just always overwriting compareFn for objects, making it impossible to pass a custom function, right?

Comment (ravishivt) [reply packages/common/src/pipes/keyvalue_pipe.ts]: Good catch!  I modified this to preserve the passed in compareFn.  I also added a test to verify that a passed in customFn works for objects.

File: packages/common/test/pipes/keyvalue_pipe_spec.ts [hunk 2]
```diff
@@ -74,27 +83,13 @@ describe('KeyValuePipe', () => {
...
       expect(pipe.transform(new Map([[1, 2]]))).toEqual([{key: 1, value: 2}]);
     });
-    it('should order by alpha', () => {
+    it('should not modify the order of keys when a compareFn is not passed', () => {
       const pipe = new KeyValuePipe(defaultKeyValueDiffers);
       expect(pipe.transform(new Map([['b', 1], ['a', 1]]))).toEqual([
-        {key: 'a', value: 1}, {key: 'b', value: 1}
-      ]);
-    });
-    it('should order by numerical', () => {
```
Comment (Toxicable) [inline packages/common/test/pipes/keyvalue_pipe_spec.ts]: By removing this test, meaning that this test is not longer correct, makes the change here breaking

Comment (ravishivt) [reply packages/common/test/pipes/keyvalue_pipe_spec.ts]: Yes, technically, this is a breaking change.  But I'd argue no-one expects the keyvalue pipe to adjust the sort order of a Map input.  Regardless, I've updated the PR description to indicate that this is indeed a breaking change.

**Comment by RobertAKARobin** [comment]
Was just about to start on a PR for this myself. Would love to see this added. Seems very strange that sorting was added to begin with. https://github.com/angular/angular/issues/31420

**Comment by RobertAKARobin** [comment]
@ravishivt Can you update this PR to pass the checks?

**Comment by ravishivt** [comment]
@RobertAKARobin I just fixed a broken test, thanks for letting me know.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
I don't know if this has been discussed already, but one way to avoid the breaking change, and also to allow JS objects to be sorted by their default ordering, would be to specify that a `compareFn` of `null` means "do not sort".

**Comment by mlc-mlapis** [comment]
> would be to specify that a compareFn of null means "do not sort".

This sounds very good and logical. Even without any breaking change.

**Comment by petebacondarwin** [comment]
I discussed this a bit with other team members and we like the idea of going with `null` as an option.

@ravishivt - I am going to close this PR since it is not the approach we want to take. If you would like to try implementing the other approach then please leave a message on the issue in the next few days. Otherwise, it will be open for others to work on.

**Pull request closed without merging**